modules/authglobus/lib/Auth/Source/Globus.php

```diff
@@ -0,0 +1,213 @@
+
+        return json_decode($info, true);
+    }
+
+    /**
+     * Obtain an Authorization Code
+     *
+     * @param array &$state  Information about the current authentication.
+     */
+    public function authenticate(&$state)
+    {
+        // We are going to need the authId in order to retrieve this authentication source later
+        $state[self::AUTHID] = $this->authId;
+
+        $stateID = SimpleSAML_Auth_State::getStateId($state);
+        SimpleSAML\Logger::debug("globus auth at state = " . $stateID);
+
+        $consumer = new sspmod_oauth_Consumer($this->key, $this->secret);
+
+        $request = array(
+            'client_id' => $this->key,
+            'scope' => $this->scope,
+            'state' => $stateID,
+            'response_type' => 'code',
+            'redirect_uri' => $this->redirect_uri
+        );
+
+        $state['globusauth:request'] = $request;
+        $urlAppend = \SimpleSAML\Utils\HTTP::addURLParameters(API_ENDPT . '/oauth2/authorize', $request);
+        SimpleSAML_Auth_State::saveState($state, self::STAGE_INIT);
+
+        // Get an Authorization Code from Globus
+        /* Example: https://auth.globus.org/v2/oauth2/authorize?client_id=69ba5e62-7285-45db-952d-e0bb73b5eac7&scope=urn:globus:auth:scope:transfer.api.globus.org:all urn:globus:auth:scope:auth.globus.org:view_identities offline_access&response_type=code&redirect_uri=https://www.example.org/my_app/login&state=g6l14b2xlgx4dtce8d2ja714i
+          */
+        $authorizeUrl = $consumer->getAuthorizeRequest($urlAppend, $request);
+    }
+
+    /**
+     * Exchange authorization code for an access token
+     *
+     * @param array &$state  Information about the current authentication.
+     */
+    public function finalStep(&$state)
+    {
+        $request = array(
+            'code' => $_GET['code'],
+            'redirect_uri' => $this->redirect_uri,
+            'grant_type' => 'authorization_code'
+        );
+        $oauthSignature = $this->key . ":" . $this->secret;
+
+        // Exchange the code we got earlier for an access token
+        $result = json_decode($this->getGlobusTokens($request, $oauthSignature));
+        $accessToken = $result->access_token;
+        $otherTokens = $result->other_tokens;
+
+        // Use this access token to tell us about our current user + affiliations
+        $userInfo = $this->getGlobusInfo(API_ENDPT . '/oauth2/userinfo', $accessToken);
+
+        // Get any other identities we may have
+        $otherIdentities = array();
+
+        if (isset($otherTokens)) {
+            foreach ($otherTokens as &$item) {
+                $res = $this->getGlobusInfo(API_ENDPT . '/api/identities/'. $userInfo['sub'], $item->access_token);
+
+                foreach ($res as &$id) {
+                    array_push($otherIdentities, $id);
+                }
+            }
+        }
+
+        $otherIdentities = $this->flatten($otherIdentities, 'identity.');
+
+        $fullname = explode(',', $userInfo['name']);
```

Comment:
are we sure that `$userInfo['name']` will always be `<last>,<first>`

What happens if it is `<last>, <first>` we should probably trim the results.

This might be better to have in the authProc as part of the config.

We can possibly also set a config to specify a "primary" identity, or find a way to use the preferred_username to find the correct identity to use.